### ISSUE TITLE
Error when calling cpp binding StartCluster

### ISSUE THREAD

**Issue by savearray2** [lni/dragonboat#33]
The following fails when using the C++ bindings:

```c++
status = nh->StartCluster(peers, false, "/Users/dict/git/dragonboat-test/dragonboat-cpp-plugin-cpphelloworld.so", config);
```

```
goroutine 17 [running, locked to thread]:
github.com/lni/dragonboat/internal/utils/fileutil.GetAppNameFromFilename(...)
	/Users/dict/.go/src/github.com/lni/dragonboat/internal/utils/fileutil/plugin.go:32
github.com/lni/dragonboat.(*NodeHost).StartClusterUsingPlugin(0xc0001a4000, 0xc0000cb200, 0x0, 0xc000612000, 0x47, 0x1, 0x80, 0x100, 0x5, 0x1, ...)
	/Users/dict/.go/src/github.com/lni/dragonboat/nodehost.go:394 +0x1b5
main.NodeHostStartCluster(0x2, 0x102a29b50, 0x102a282d0, 0x3, 0x0, 0x102a28d70, 0x47, 0x1, 0x80, 0x50000100, ...)
	/Users/dict/.go/src/github.com/lni/dragonboat/binding/binding.go:291 +0x395
main._cgoexpwrap_2a7f7706cd22_NodeHostStartCluster(0x2, 0x102a29b50, 0x102a282d0, 0x3, 0xa00000000, 0x102a28d70, 0x47, 0x1, 0x80, 0x50000100, ...)
	_cgo_gotypes.go:369 +0xad
```

It works when a relative path is used instead:

```c++
status = nh->StartCluster(peers, false, "dragonboat-cpp-plugin-cpphelloworld.so", config);
```

The problem seems to originate here:
https://github.com/lni/dragonboat/blob/ca03e0312b1dd2958fd0c618a9c42257d3702e6e/internal/utils/fileutil/plugin.go#L23-L33

A potential fix would be to remove the `^` from the beginning of the applicable regular expression(s), or change it to `.*`.

But it may be best if we could set our own app name, rather than forcing dynamically created libraries to have a certain path and/or name.

Thanks for your work.

**Comment by lni** [comment]
@savearray2 Thanks for the feedback. I agree that it will be more convenient to also allow users to just pass in a C++ factory function for creating the StateMachine. This will also make the C++ API more consistent with the native Go API. 

Will probably do it next week, as I am busy with a long list of new features/refactoring.

**Comment by savearray2** [comment]
@lni No problem. I appreciate your response. Right now I'm just using a symbolic link that points back to my compiled executable as a stopgap. For better or worse, it seems to work fine for now.

Also, while you're on the topic of refactoring, what do you believe is the best way to store a large amount of key/value pairs on each node sent through proposals to a dragonboat cluster (1-100k each message). I'm planning on implementing a simple sharding/balancing system, but it's possible that's something dragonboat already has in some form.

Should I just run a separate RocksDB instance alongside the WAL's that dragonboat produces? Somehow this seems potentially inefficient. I believe I read that there's an easy way to intercept requests sent to the DB layer by utilizing an interface, but I have yet to have a chance to dig too deeply into the code.

Also, do you know generally what features you'll be adding? I'll probably hold off on any heavy development if you're only a week or two out from completion.

I really appreciate your work here. The library is great and seems solid so far.

**Comment by lni** [comment]
@savearray2 It really depends on how many such Key-Value pairs you are going to deal with. If the amount is a couple million pairs and the total size can be comfortably fitted into your RAM, then you can even store all key-value pairs in an in-memory map. Note that all your data will still be persisted as all proposals are stored on disk. 

If your data size is larger than your RAM, then a RocksDB style disk based KV store is probably better. I'd recommend to use the recently added IConcurrentStateMachine SM to allow concurrent access to the underlying data store. It allows you to concurrently query your Key-Value pairs when your proposals are being applied or snapshots are being created. 

That being said, however, the C++ binding have not been updated to cover the IConcurrentStateMachine support yet. It is a small update as it is just some additional wrapper code. Are you going to use dragonboat as a native Go library or use it in your C++ project? 

The new features being developed are detailed in #34 and #35. Basically, when some projects choose to have their state machine data stored on disk, the new feature will allow them to create and stream the snapshot to a remote node on the fly without periodically generating snapshots in the background. It saves disk space and bandwidth by eliminating redundant snapshots written to disks, it is more like some optimizations.

**Comment by savearray2** [comment]
@lni I'm dealing with mostly commodity hardware, so each node probably won't have more than 8-16 GB of RAM. I might be able to stretch it to 32 GB, if I have to. Everything will run ECC RAM, however. Additionally, new nodes will come online at random (without being able to predict when) from week to week, with other nodes being removed week to week (but with more control when that happens). Each node is expected to hold anywhere between 1-3 TB of data. Data will be continuously added to the cluster, but at very slow rates (it's likely that update operations will continuously hit the cluster, however). Reads will also be somewhat rare, but I do have the requirement of linearizability, atomic operations, and that a majority has recorded a given key/value pair (consensus), with reads being fully fault tolerance and highly available. There will also be a requirement to add a TTL to some of the data, so that it will automatically be erased after a certain period of time. So I'll have to implement my own sharding/balancing system, I imagine. I believe I can just use the ClusterID that dragonboat provides to do the sharding, which is one of the things that drew me to dragonboat, and the balancing can take place as a journaled operation on top of the cluster. Effectively, the system needs to be set up to handle an unlimited amount of key/value data storage.

It does sound like the IConcurrentStateMachine interface is the way to go. I suppose I can write my own add-on/patch to implement any interfaces that I need, or write certain parts of the system in pure Go. I've been working with a lot of different languages for this project so far, but C++ and Rust are the largest two. I actually know very little Go, but I don't think it should be that hard to manipulate with some C knowledge. I was hoping to plug directly into the C++ binding to avoid Go (also if I'm already using C++, I might as well try to avoid the GC, as well), but I'm open to options, especially if they work well.

---
 
#34 & #35 sounds pretty interesting, as you could implement snapshots using technologies like zfs & possibly even zfs send/receive. This would actually fit my use case extremely well. It would also allow for extremely space-efficient snapshots, as they would be differential backups instead of full backups.

If there was a notification (callback) sent that a snapshot needed to be made, as most databases need to shut down/pause connections/flush to disk, the implementation could handle shutting down the DB, and then perform the zfs snapshot. Then, if a system needed to be recovered, it could possibly restore the data from a hard backup somewhere, and then perform further syncing via a zfs send/receive, and then finally use dragonboat to catch up with the rest of the cluster. A rsync could even be used to handle making a snapshot. We could also add in additional logic to ensure that a snapshot is byte-per-byte accurate after it has been transmitted to an off-site location.

It would also be extremely helpful to allow certain nodes to make snapshots while others are exempt from the snapshot process. If I have to snapshot 3 TB of data to a backup disk, or even to a remote file store, like Amazon or S3 Glacier, the snapshotting process would most likely take a while. It would also be nice if we could rotate the snapshot duties such that we can have one snapshot being made at a time, or manual/granular control over the snapshotting process so that we can control them dynamically through code. In my case, I most likely will archive and then encrypt our file backups, and upload them to Amazon's S3 Glacier (for cheap long term storage) on the fly, all through a single stream in memory.

Also, at least in my case, it's somewhat common to go back and increase the size of a disk/partition, so it is possible that the underlying storage size changes. Hopefully dragonboat will take this into account.

Thanks again for your help, and sorry for the long post. Hopefully I was able to be helpful with some of my comments here.

**Comment by lni** [comment]
@savearray2 thanks for the detailed descriptions

> Each node is expected to hold anywhere between 1-3 TB of data.

When there is terabytes of data on each machine, it is quite often to further shard it into many Raft groups. For example, if each Raft group looks after say 1Gbytes data, there will be 1,024 Raft groups for each 1T data. This avoids the issue that multi-terabytes might need to be snapshotted in one go.

> I was hoping to plug directly into the C++ binding to avoid Go

As the library is implemented in Go, I hope you can understand that the C++ binding support will probably be a bit behind in terms of new features & fixes. I do hope people might find the C++ binding useful and start contributing code for it. ;)

Personally, I found Go very easy to learn and use. I used to worry about its GC performance, but it turned out to be pretty impressive. 

> the implementation could handle shutting down the DB, and then perform the zfs snapshot

Dragonboat doesn't assume to be running on any particular filesystem, so it doesn't use any filesystem specific feature such as zfs snapshot. 

In the current implementation I am working on, a snapshot is required when certain followers are left far behind (e.g. too slow or a new machine just joined). The State Machine on the leader will be notified and a callback method provided by user application will be invoked to generate such snapshot. It requires to be a concurrent snapshot meaning both read/write operations should still be possible when the snapshot is being generated. When using RocksDB to manage application data, user can use the snapshot feature provided by RocksDB to concurrently generate such a point-in-time dump of all data.

> In my case, I most likely will archive and then encrypt our file backups

When using Raft, there will be multiple copies of your data by default. Losing a minority of those copies won't affect the availability of your system/data and thus you don't need to deal with backups in such case. You can just join some new nodes to replace the failed ones, once they catch up with the existing nodes then everything will be fine. 

Manually restoring a system using saved snapshot images will only help when the majority of those copies are gone. In such case, like most other consensus based libraries, the restore procedures will be more complicated than just "copy all saved files from last night". I will release some tools/docs to help users to do that. It is on the top of my list when #34 and #35 are released as beta. 

I will write some wiki articles in the next couple of weeks on some DevOps related stuff, hopefully that can address some common questions. You covered a few of them in your post, I guess many potential users are going to face similar ones, I am happy to give more details.

**Comment by savearray2** [comment]
@lni Thank you for the detailed explanation. Splitting the nodes up to handle much smaller amounts of data, but possibly running multiple on the same system is a very good idea. The snapshots taken will also be much more randomized across the cluster.

However, I still think it would be a good idea for the user to be able to take somewhat manual control over when each node snapshots. There's a requirement for my specific needs that forces me to do geographically-distributed cold storage off-site for extreme _disaster recovery_ and _auditing purposes_, so it cannot be a constant connection, and I have to take snapshots of different portions of the data at different times. Maybe I could send a proposal across the cluster that would request a certain node make a backup. Basically, an additional protocol on top of a dragonboat cluster. The problem isn't unworkable in the least.

Other than that, everything you've said makes perfect sense, and I appreciate the work you're doing. I know it's a pain to maintain something in a quality sufficient to release to the public.

Thanks again :)

**Comment by lni** [comment]
@savearray2 

I guess my previous answer caused some confusions. Let me try again  ;)

In the current released version v2.1.3, users can't manually request a snapshot to be taken immediately. Users can only specify how often (in terms of how many new updates) snapshots should be captured & saved. Please also note that the term "snapshot" here is a point-in-time view of the State Machine described in the raft paper, not something users generate in their applications. 

I am working on a major release v3.0 with many new major features - all new features have been implemented, it is being actively tested. Once that is released, in the v3.1, the biggest change will be the manual snapshot feature to allow users to trigger snapshot to be saved. 

What I am trying to point out is that there is a well defined procedure to restore a permanently damaged (meaning the majority of nodes are permanently gone/dead) raft group from such saved snapshot. Because of that, new tools/docs will be provided as a part of the v3.1 release.
